extensions/stats/plugin.cc

```diff
@@ -84,42 +95,109 @@ bool PluginRootContext::onConfigure(size_t) {
+          MetricType::Counter,
+          [](const ::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+            return request_info.tcp_received_bytes;
+          },
+          field_separator, value_separator, true),
+      StatGen(
+          absl::StrCat(stat_prefix, "tcp_connections_opened_total"),
+          MetricType::Counter,
+          [](const ::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+            return request_info.tcp_connections_opened;
+          },
+          field_separator, value_separator, true),
+      StatGen(
+          absl::StrCat(stat_prefix, "tcp_connections_closed_total"),
+          MetricType::Counter,
+          [](const ::Wasm::Common::RequestInfo& request_info) -> uint64_t {
+            return request_info.tcp_connections_closed;
+          },
+          field_separator, value_separator, true),
   };
+
+  long long tcp_report_duration_nanos = 0;
+  if (config_.has_tcp_reporting_duration()) {
+    tcp_report_duration_nanos =
+        ::google::protobuf::util::TimeUtil::DurationToNanoseconds(
+            config_.tcp_reporting_duration());
+  } else {
+    tcp_report_duration_nanos = kDefaultTCPReportDurationNanoseconds;
+  }
+  proxy_set_tick_period_milliseconds(tcp_report_duration_nanos);
   return true;
 }
 
-void PluginRootContext::report(bool is_tcp) {
-  const auto peer_node_ptr =
-      node_info_cache_.getPeerById(peer_metadata_id_key_, peer_metadata_key_);
+void PluginRootContext::onTick() {
+  if (tcp_request_queue_.size() < 1) {
+    return;
+  }
+  for (auto const& item : tcp_request_queue_) {
+    // requestinfo is null, so continue.
+    if (item.second == nullptr) {
+      continue;
+    }
+    if (report(*item.second, true)) {
+      // Clear existing data in TCP metrics, so that we don't double count the
+      // metrics.
+      clearTcpMetrics(*item.second);
+    }
+  }
+}
+
+bool PluginRootContext::report(::Wasm::Common::RequestInfo& request_info,
+                               bool is_tcp) {
+  std::string peer_id;
+  const auto peer_node_ptr = node_info_cache_.getPeerById(
+      peer_metadata_id_key_, peer_metadata_key_, peer_id);
+  // For TCP, if metadata exchange is not available, peer id is set as peer
+  // metadata id key. Otherwise, metadata exchange has not happened yet and we
+  // wait for it to happen before we report any metric.
+  // TODO(gargnupur): Remove outbound_ from condition below, when
+  // https://github.com/envoyproxy/envoy-wasm/issues/291 is fixed.
+  if (is_tcp && peer_node_ptr == nullptr && peer_id != peer_metadata_id_key_ &&
+      !outbound_) {
+    return false;
+  }
   const wasm::common::NodeInfo& peer_node =
       peer_node_ptr ? *peer_node_ptr : ::Wasm::Common::EmptyNodeInfo;
 
   // map and overwrite previous mapping.
   const auto& destination_node_info = outbound_ ? peer_node : local_node_info_;
-  ::Wasm::Common::RequestInfo request_info;
   if (is_tcp) {
     ::Wasm::Common::populateTCPRequestInfo(outbound_, &request_info,
```

Comment:
This should not be called on every report call right? It should only be called once when tcp peer metadata is found.

Reply:
good point.. will fix